Add patient sidebar menu options for patient status

nmajor25 opened the pull request:
Shortcut Story ID: [sc-46925]

Adds two new options to the patient sidebar menu to allow a user to update a patient's status:

1. Toggle between `active` & `inactive` status.
2. Set a patient as `archived`.

This PR also implements the `patients:manage` permission. These menu options won't be shown to users that don't have that permission.

Still yet to do:

- [x] Refactoring
- [x] Add confirmation modal before setting archived status
- [x] Add/update tests
- [x] Update language in menu & archive confirmation modal

nmajor25 committed:
Add menu options to patient sidebar for toggling active/inactive patient status and setting a patient as archived

nmajor25 committed:
Handle 410 error when retrieving patient from api

After a patient is archived in a specific workspace, that patient will return a 410

nmajor25 commented on src/js/apps/patients/patient/sidebar/sidebar_app.js:
```diff
@@ -25,15 +28,29 @@ export default App.extend({
   onStart({ patient }) {
     this.patient = patient;
+    const currentUser = Radio.request('bootstrap', 'currentUser');
+    this.canManagePatients = currentUser.can('patients:manage');
```
Based on what's in the backend, `patients:manage` is needed in order to update the patient's status. And the api request fails if the user doesn't have it.

There is also a `patients:owned:manage` permission in the backend. But it doesn't look like it's used for anything.

nmajor25 commented on src/js/entities-service/entities/workspace-patients.js:
```diff
@@ -6,6 +7,19 @@ const TYPE = 'workspace-patients';
+    };
+
+    this.save(attrs, { relationships }, opts);
+  },
```
A lot of this probably needs to be refactored. This is in the "get it to work" state.

paulfalgout replied:
Let's make this function a `saveAll` and instead of `newStatus` just pass in `{ status: 'foo' }`

paulfalgout replied:
Rather than needing to pass the patient here, let's modify the initial entity service request to do:
```js
    return new Model({ id: uuid(patientId, workspaceId), _patient: patientId, _workspace: workspaceId });
```
so that we do not need to re-get the workspace or pass the patient.. the data should always match.

nmajor25 replied:
Good call, updated to that 👍

nmajor25 commented on src/js/views/patients/patient/sidebar/sidebar_views.js:
```diff
@@ -61,21 +61,35 @@ const SidebarView = View.extend({
+        menuOptions.push({
+          text: i18n.menuOptions.archive,
+          onSelect: bind(this.triggerMethod, this, 'click:archivedStatus'),
+        });
```
Archiving a patient seems like a big deal. And something that can't be easily reversed in most cases.

Should we add a `Are you sure?` confirmation modal? Could also help users avoid setting `status = archived` on accident.

paulfalgout replied:
Definitely could use a modal here

nmajor25 replied:
Confirmation modal has been added:


https://github.com/RoundingWell/care-ops-frontend/assets/35355575/2a2b04b7-896e-41c7-b246-09fbb42c2c81

nmajor25 commented on src/js/apps/patients/patient/sidebar/sidebar_app.js:
```diff
@@ -25,15 +28,29 @@ export default App.extend({
   onStart({ patient }) {
     this.patient = patient;
+    const currentUser = Radio.request('bootstrap', 'currentUser');
+    this.canManagePatients = currentUser.can('patients:manage');
```
When a patient is archived, that patient should be removed from patient search results, correct? (in that specific workspace)

In my local sandbox, the patients are still being included. Not sure if that translates to production, but wanted to make a note of that.

paulfalgout replied:
I don't know on this one.. I would assume not, but not sure

nmajor25 replied:
In that scenario, if a user clicks on an archived patient in the search results, the user will hit that 410/404 error page.

nmajor25 commented on src/js/apps/patients/patient/sidebar/sidebar_app.js:
```diff
@@ -25,15 +28,29 @@ export default App.extend({
   },
+  toggleActiveStatus() {
+    this.patient.toggleActiveStatus();
+    this.showSidebar();
```
When the status of the patient is changed, the text option in the menu and the status widget will need to re-render to show the updated status value.

Not sure how to re-render those individually... so I'm just re-rendering the entire sidebar to get those values to update. Seems okay to me, but open to a better way to do it.

paulfalgout replied:
So on the status widget we can add an `initialize` where we get the `workspacePatient` and then just listenTo the `change:status` on that and re-render.

nmajor25 replied:
That worked nicely 👍

nmajor25 commented on src/js/apps/patients/patient/patient_app.js:
```diff
@@ -42,6 +42,11 @@ export default SubRouterApp.extend({
+    Radio.trigger('event-router', 'notFound');
+    this.stop();
+  },
+
```
After a patient is archived, retrieving that patient returns a `410` error from the api. (for that specific workspace)

In that scenario, this shows the `404` page error. Not sure what we want to show.

<img width="347" alt="Screenshot 2024-01-26 at 2 36 34 PM" src="https://github.com/RoundingWell/care-ops-frontend/assets/35355575/1c7ba375-def7-4999-8503-faf3259b7e1f">

paulfalgout replied:
Yep. so with our api 404 means that api endpoint doesn't exist.  410 means it's not found on an existing api endpoint.  So this is what we want to show

paulfalgout replied:
could do: `onFail(options, { response }) { if (response.status === 410`

nmajor25 replied:
Updated this to do the `if (response.status === 410)`.

paulfalgout commented on src/js/views/patients/patient/sidebar/sidebar_views.js:
```diff
@@ -61,21 +61,35 @@ const SidebarView = View.extend({
+        onSelect: bind(this.triggerMethod, this, 'click:activeStatus'),
+      });
+
+      if (patientStatus !== 'archived') {
```
can we move these strings into the static.js?

nmajor25 replied:
Done 👍

paulfalgout commented on src/js/views/patients/patient/sidebar/sidebar_views.js:
```diff
@@ -61,21 +61,35 @@ const SidebarView = View.extend({
       },
     ]);
 
+    if (canEdit && canManagePatients) {
```
is this really `canEdit`?  Can only manually added patients status be adjusted?

I wonder if really this check should be `this.model.getWorkspacePatient().canEdit()`

nmajor25 replied:
Hmm right, probably don't need to check if it's a manually added patient.

In https://github.com/RoundingWell/care-ops-frontend/pull/1227/commits/11b4960b624c3b100536e0d2f743af89e3773186, I created a `workspacePatient.canEdit()` function and use that here instead. Where I check `isNew()` and the `currentUser.can('patient:manage')` permission.

paulfalgout replied:
I guess I'm not sure if we need the isNew check.  I don't think this sidebar can be shown for a new patient, so we'll never be able to hit that coverage

nmajor25 replied:
Okay, I removed the `isNew()` check from `workspacePatient.canEdit()`. Which now only checks that the current user has the `patient:manage` permission.

paulfalgout commented on src/js/entities-service/entities/patients.js:
```diff
@@ -52,6 +52,18 @@ const _Model = BaseModel.extend({
     return workspacePatient.get('status');
   },
+  toggleActiveStatus() {
+    const workspacePatient = Radio.request('entities', 'get:workspacePatients:model', this.id);
```
Let's remove the `getStatus` function and just do `getWorkspacePatient` for a single way to look it up on the patient

nmajor25 replied:
Done 👍

paulfalgout replied:
I mean can we replace the `getStatus()` function from this entity up on 53 and make a `getWorkspacePatient`?

nmajor25 replied:
I believe that's the update I made... this `/entities/patients.js` file no longer has the `getStatus()` function. And I created the `getWorkspacePatient()` and use that instead of `patient.getStatus()`.

nmajor25 committed:
Refactor workspace-patients entities 'setNewStatus()' method to a 'saveAll()'

nmajor25 committed:
Rerender status widget in patient sidebar when workspace-patient status is changed

Do this instead of rerendering the entire patient sidebar

nmajor25 committed:
Only show patients app onFail() error if 'response.status === 410'

nmajor25 committed:
Move workspace patient status strings to /src/js/static.js file

nmajor25 commented on src/js/views/patients/widgets/widgets.js:
```diff
@@ -64,14 +64,21 @@ const widgets = {
     template: hbs`{{formatMessage (intlGet "patients.widgets.widgets.sex") sex=sex}}`,
   },
-  status: {
+  status: View.extend({
```
Had to do `View.extend` here to get the `initialize()` function to work.

nmajor25 committed:
Replace `patient.getStatus()` with `patient.getWorkspacePatient()` that can be used universally to get the workspace-patient status

nmajor25 committed:
Add `workspacePatient.canEdit()` function and use it for determining if the update status options in the patient sidebar menu should show

nmajor25 commented on src/js/entities-service/entities/workspace-patients.js:
```diff
@@ -16,6 +17,12 @@ const _Model = BaseModel.extend({
+    const patient = Radio.request('entities', 'patients:model', this.get('_patient'));
+    const currentUser = Radio.request('bootstrap', 'currentUser');
+
+    return patient.isNew() || currentUser.can('patients:manage');
```
Do we need the `isNew()` check here? Not 100% sure...

paulfalgout commented on src/js/views/patients/widgets/widgets.js:
```diff
@@ -64,14 +64,21 @@ const widgets = {
+  status: View.extend({
     template: hbs`<span class="widgets__status-{{ status }}">{{formatMessage (intlGet "patients.widgets.widgets.status") status=status}}</span>`,
+    initialize() {
+      const workspacePatient = Radio.request('entities', 'get:workspacePatients:model', this.model.get('id'));
```
never have to do `.get('id')` `.id` is always better.  But instead can this call `this.model.getWorkspacePatient()`?  We can also probably just `this.listenTo(workspacePatient, 'change:status', this.render)`

nmajor25 replied:
I updated the listener to this in the status widget:

```js
initialize() {
  this.workspacePatient = this.model.getWorkspacePatient();

  this.listenTo(this.workspacePatient, 'change:status', this.render);
},
templateContext() {
  return {
    status: this.workspacePatient.get('status'),
  };
},
```

nmajor25 committed:
Refactor status widget change listener code

nmajor25 committed:
Show a confirmation modal before setting a patient's status to `'archived'`.

nmajor25 committed:
Add test for updating patient status via patient sidebar menu

nmajor25 committed:
Add test for 410 error on patient dashboard

After a patient is archived, the api request for that patient will return a 410 error

nmajor25 committed:
Update some language in the patient sidebar menu

nmajor25 committed:
Update language in archive patient confirmation modal

nmajor25 committed:
Add `.wait('@routeWorkspacePatient')` to `update patient status` patient sidebar test
